gapic/src/main/com/google/gapid/models/Devices.java

```diff
@@ -142,6 +144,48 @@ public void selectReplayDevice(Device.Instance dev) {
     listeners.fire().onReplayDeviceChanged(dev);
   }
 
+  public void updateValidationStatus(DeviceCaptureInfo device, Service.ValidateDeviceResponse response) {
+    if (response == null || response.hasError()) {
+      device.validationStatus = false;
+      return;
+    }
+    device.validationStatus = true;
+  }
+
+  public void validateDevice(DeviceCaptureInfo device, Runnable callback) {
+    if (!devices.contains(device)) {
+      return;
+    }
+    rpcController.start().listen(client.validateDevice(device.path),
+        new UiErrorCallback<Service.ValidateDeviceResponse, Service.ValidateDeviceResponse, Service.ValidateDeviceResponse>(shell, LOG) {
+
+      @Override
+      protected void onUiThreadSuccess(Service.ValidateDeviceResponse response) {
+        updateValidationStatus(device, response);
+        callback.run();
+      }
+
+      @Override
+      protected void onUiThreadError(Service.ValidateDeviceResponse response) {
+        LOG.log(Level.WARNING, response.toString());
+        updateValidationStatus(device, response);
+        callback.run();
+      }
+
+      @Override
+      protected ResultOrError<ValidateDeviceResponse, ValidateDeviceResponse> onRpcThread(Rpc.Result<ValidateDeviceResponse> response)
+          throws RpcException, ExecutionException {
+        try {
+          return success(response.get());
+        } catch (RpcException | ExecutionException e) {
+          throttleLogRpcError(LOG, "LoadData error", e);
+          return error(null);
+        }
+      }
```

Comment:
Please move this above the `onUiThread` methods, because it is run before them. Make it easier to read.

Reply:
Done.